### PR TITLE
Add option to notify when synced folder overcomes set size limit

### LINKED ISSUE #5818: Warn if folder during sync overcomes configured threshold
In Settings we have
- confirm to sync folder if it is bigger than: [ threshold ]

This does work only for initial sync:
- client checks on first sync if folder is bigger than threshold

It will not work in this case:
- threshold is 500Mb
- folder A is empty
- it will be synced
- user A adds 600Mb file into file A

So our idea is, to stay backward compatible:
- add sub setting "obey also during sync"
- this then checks on each and every sync if folder on server is above threshold
- if so:
  - ask user via activity tray "Folder A will overcome threshold. What do you" "sync" vs "stop sync"?
  - if user choose "sync", store this info in local db, so that even if folder shrinks and reaches limit again, user will not be asked
  - if user choose "no", selective sync will be dropped and folder will be deleted locally (like it happens already now, if doen manually)
- with existing folders same logic applies
- when user reduces threshold, show a warning that this might affect existing folders, then apply above logic

### PR DESCRIPTION
**Pull request by claucambra**
This expands upon the existing feature to not sync new folders larger than a set limit

Closes #5818 

Screenshot 2023-07-05 at 23 40 28

Screenshot 2023-07-06 at 14 46 01

### CODE REVIEW

**Comment by mgallien** [comment]
@claucambra build is failing

**Comment by claucambra** [comment]
> @claucambra build is failing

Now fixed

**Comment by tobiasKaminsky** [comment]
@claucambra I gave this a test run 
- enabled above option, limit to 50Mb
- create folder on server
- added 20kb
- synced
- added 54Mb
--> file was downloaded without problems

It is shown here:
image

**Comment by claucambra** [comment]
> @claucambra I gave this a test run
> 
>     * enabled above option, limit to 50Mb
> 
>     * create folder on server
> 
>     * added 20kb
> 
>     * synced
> 
>     * added 54Mb
>       --> file was downloaded without problems
> 
> 
> It is shown here: image

Great, working as expected then :) did you get the notification?

**Comment by tobiasKaminsky** [comment]
No, no notification.
And it should not sync folder, but first ask what to do, like in your first screenshot:
- client notices folder about limit
- client pauses sync for now
- client asks user what to do
- "keep syncing" -> resume sync 
- "stop syncing" -> remove folder locally

**Review by mgallien — CHANGES_REQUESTED** [review]
@claucambra please add automated tests covering teh changes in libsync code (mainly `SyncEngine` and related classes

**Comment by claucambra** [comment]
> No, no notification. And it should not sync folder, but first ask what to do, like in your first screenshot:
> 
>     * client notices folder about limit
> 
>     * client pauses sync for now
> 
>     * client asks user what to do
> 
>     * "keep syncing" -> resume sync
> 
>     * "stop syncing" -> remove folder locally

As far as I understood the task is to *warn* the user, not stop syncing the folder/stop the current sync run -- I imagine this would be annoying if sync suddenly stopped when folders grow, no?

Could you also send client logs for the sync run to debug the lack of a notification? Thanks!

**Comment by tobiasKaminsky** [comment]
As the other options are also stated "ask for confirmation" I think we should do the same here.
I am sorry that I was not clear enough when writing description.

Once you have changed it, I can re-test and send debug archive if needed.

**Comment by claucambra** [comment]
> As the other options are also stated "ask for confirmation" I think we should do the same here. I am sorry that I was not clear enough when writing description.
> 
> Once you have changed it, I can re-test and send debug archive if needed.

I have added an additional checkbox to disable synchronisation of files that have overcome limit now, feel free to test :)

**Comment by tobiasKaminsky** [comment]
Works great!
Ready for merge :-)

File: src/gui/folder.cpp [hunk 1]
```diff
@@ -1564,11 +1673,8 @@ bool FolderDefinition::load(QSettings &settings, const QString &alias,
 
 QString FolderDefinition::prepareLocalPath(const QString &path)
 {
-    QString p = QDir::fromNativeSeparators(path);
-    if (!p.endsWith(QLatin1Char('/'))) {
-        p.append(QLatin1Char('/'));
-    }
-    return p;
+    const auto normalisedPath = QDir::fromNativeSeparators(path);
```
[the remark this answers is not in the document]
Comment (claucambra) [reply src/gui/folder.cpp:1676]: not a typo ;)

File: src/libsync/discoveryphase.cpp [hunk 2]
```diff
@@ -103,41 +111,38 @@ void DiscoveryPhase::checkSelectiveSyncNewFolder(const QString &path, RemotePerm
...
-            if (!p.endsWith(QLatin1Char('/')))
-                p += QLatin1Char('/');
-            _selectiveSyncWhiteList.insert(
-                std::upper_bound(_selectiveSyncWhiteList.begin(), _selectiveSyncWhiteList.end(), p),
-                p);
-            return callback(false);
+        }
+
+        // it is not too big, put it in the white list (so we will not do more query for the children) and and do not block.
+        const auto sanitisedPath = Utility::trailingSlashPath(path);
```
[the remark this answers is not in the document]
Comment (claucambra) [reply src/libsync/discoveryphase.cpp:126]: not a typo ;)

File: src/gui/folder.cpp [hunk 3]
```diff
@@ -1212,7 +1242,86 @@ void Folder::slotNewBigFolderDiscovered(const QString &newF, bool isExternal)
...
+            blacklistActivityLink._verb = "BLACKLIST_FOLDER";
+
+            activityLinks.append(blacklistActivityLink);
+        }
+
+        auto existingFolderNowBigActivity = Activity();
+        existingFolderNowBigActivity._type = Activity::NotificationType;
+        existingFolderNowBigActivity._dateTime = QDateTime::fromString(QDateTime::currentDateTime().toString(), Qt::ISODate);
+        existingFolderNowBigActivity._subject =
+            tr("The folder %1 has surpassed the set folder size limit of %2MB.").arg(folderPath, QString::number(ConfigFile().newBigFolderSizeLimit().second));
```
[the remark this answers is not in the document]
Comment (claucambra) [reply src/gui/folder.cpp:1315]: done

File: src/gui/tray/activitylistmodel.cpp [hunk 4]
```diff
@@ -848,6 +848,14 @@ void ActivityListModel::slotTriggerAction(const int activityIndex, const int act
                (activity._syncFileItemStatus == SyncFileItem::Conflict || activity._syncFileItemStatus == SyncFileItem::FileNameClash)) {
         slotTriggerDefaultAction(activityIndex);
         return;
+    } else if (action._verb == "WHITELIST_FOLDER" && !activity._file.isEmpty()) { // _folder == folder alias/name, _file == folder/file path
```
[the remark this answers is not in the document]
Comment (claucambra) [reply src/gui/tray/activitylistmodel.cpp:851]: done

File: src/gui/folder.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (claucambra) [reply src/gui/folder.cpp:1315]: Actually no, this generates a warning about losing precising as arg converts to int -- will retain `QString::number`

File: src/gui/folder.cpp [hunk 5]
```diff
@@ -1212,7 +1242,86 @@ void Folder::slotNewBigFolderDiscovered(const QString &newF, bool isExternal)
...
+    auto undecidedListQueryOk = false;
+    auto undecidedList = journal->getSelectiveSyncList(SyncJournalDb::SelectiveSyncUndecidedList, &undecidedListQueryOk);
+    if (undecidedListQueryOk) {
+        if (!undecidedList.contains(trailSlashFolderPath)) {
+            undecidedList.append(trailSlashFolderPath);
+            journal->setSelectiveSyncList(SyncJournalDb::SelectiveSyncUndecidedList, undecidedList);
+            emit newBigFolderDiscovered(trailSlashFolderPath);
+        }
+
+        const auto messageInstruction =
```
[the remark this answers is not in the document]
Comment (claucambra) [reply src/gui/folder.cpp]: Done now

File: src/gui/folderman.cpp [hunk 6]
```diff
@@ -1259,6 +1259,30 @@ Folder *FolderMan::folderForPath(const QString &path)
...
+void FolderMan::blacklistFolderPath(const QString &path)
+{
+    const auto folder = folderForPath(path);
+    if (!folder) {
+        return;
+    }
+
+    const QString folderPath = folder->cleanPath() + QLatin1Char('/');
+    const auto relPath = path.mid(folderPath.length());
+    folder->blacklistPath(relPath);
```
[the remark this answers is not in the document]
Comment (claucambra) [reply src/gui/folderman.cpp]: Done now

**Review by allexzander — APPROVED** [review]

**Pull request merged**